File: lib/ai/schemas.ts
import { z } from "zod";

import { customFieldTypeValues, issuePriorityValues } from "@/lib/db/schema";

// A detected value for one of the space's existing custom fields, keyed by
// the field's `key` (not id — the route resolves key -> id, same spirit as
// repo/milestone name resolution). `value` is always a string; the route's
// resolveAiCustomFields coerces it per the field's real type (number/date
// parsing, option-label matching) since a fixed JSON schema can't vary its
// value type per dynamic field.
const aiFieldValueSchema = z.object({
  fieldKey: z.string().min(1).max(60),
  value: z.string().min(1).max(200),
});

// A field the AI thinks should exist but doesn't yet (e.g. "Size: L" with no
// Size field in the space) — the route creates it via createCustomField.
const aiNewFieldSchema = z.object({
  key: z.string().min(1).max(60),
  name: z.string().min(1).max(60),
  type: z.enum(customFieldTypeValues),
  options: z.array(z.string().min(1).max(60)).max(10).optional(),
});

// Shape returned by the "paste a task" AI flow: one raw text blob in,
// one structured task draft out. Validated even though we also ask Claude
// for constrained/JSON output — structured output isn't a 100% guarantee.
export const parsedTaskSchema = z.object({
  title: z.string().min(1).max(200),
  description: z.string().max(2000).optional(),
  priority: z.enum(issuePriorityValues),
  tags: z.array(z.string().min(1).max(30)).max(8).default([]),
  branch: z.string().max(100).optional(),
  estimate: z.string().max(20).optional(),
  dueDate: z.string().optional(), // ISO date (YYYY-MM-DD), if Claude could resolve one
  repos: z.array(z.string().max(100)).max(5).optional(), // repo names (not ids) — resolved to repoIds by the route
  milestone: z.string().max(150).optional(), // milestone title (not id) — resolved to milestoneId by the route
  fieldValues: z.array(aiFieldValueSchema).max(10).optional(),
  newFields: z.array(aiNewFieldSchema).max(3).optional(),
});
export type ParsedTask = z.infer<typeof parsedTaskSchema>;

// Shape returned by the "paste a list" bulk AI flow: one raw text blob in
// (a bullet/numbered/one-per-line list), one structured task per item out.
export const parsedTasksSchema = z.array(parsedTaskSchema).min(1).max(50);
export type ParsedTasks = z.infer<typeof parsedTasksSchema>;

export const attachedImageSchema = z.object({
  mimeType: z.string().min(1),
  data: z.string().min(1), // raw base64, no "data:...;base64," prefix
});
export type AttachedImage = z.infer<typeof attachedImageSchema>;

export const parseTaskRequestSchema = z
  .object({
    text: z.string().max(4000).default(""),
    spaceId: z.string().uuid().optional(),
    attachedImage: attachedImageSchema.optional(),
  })
  .refine((data) => data.text.trim().length >= 3 || data.attachedImage, {
    message: "Paste some text or attach a screenshot.",
  });

export const parseTasksRequestSchema = z
  .object({
    text: z.string().max(8000).default(""),
    spaceId: z.string().uuid().optional(),
  })
  .refine((data) => data.text.trim().length >= 3, {
    message: "Paste a list of tasks.",
  });

// Shape returned by the "Generate subtasks" AI flow in the task detail drawer.
export const generatedSubtasksSchema = z.array(z.string().min(1).max(200)).min(1).max(10);

export const generateSubtasksRequestSchema = z.object({
  title: z.string().trim().min(1).max(200),
  description: z.string().trim().max(5000).optional(),
});

File: Packages/ZenithAI/Sources/ZenithAI/Schemas.swift
import Foundation
import ZenithData

/// Port of `lib/ai/schemas.ts` — shapes returned by the AI parse flows,
/// validated after decoding since the CLI's structured output isn't a 100%
/// guarantee (same reasoning as the TS side re-validating with zod even
/// though the request already asked for JSON-schema-constrained output).

public struct AttachedImage: Codable, Sendable, Equatable {
    public var mimeType: String
    /// Raw base64, no `data:...;base64,` prefix.
    public var data: String

    public init(mimeType: String, data: String) {
        self.mimeType = mimeType
        self.data = data
    }
}

/// A detected value for one of the space's existing custom fields, keyed by
/// the field's `key` (not id) — `AIOrchestration` resolves key -> id.
public struct AIFieldValue: Codable, Sendable, Equatable {
    public var fieldKey: String
    public var value: String
}

/// A field the AI thinks should exist but doesn't yet.
public struct AINewField: Codable, Sendable, Equatable {
    public var key: String
    public var name: String
    public var type: CustomFieldType
    public var options: [String]?
}

public struct ParsedTask: Codable, Sendable, Equatable {
    public var title: String
    public var description: String?
    public var priority: IssuePriority
    public var tags: [String]
    public var branch: String?
    public var estimate: String?
    /// ISO date (YYYY-MM-DD), if Claude could resolve one.
    public var dueDate: String?
    /// Repo names (not ids) — resolved to repoIds by `AIOrchestration`.
    public var repos: [String]?
    /// Milestone title (not id) — resolved to milestoneId by `AIOrchestration`.
    public var milestone: String?
    public var fieldValues: [AIFieldValue]?
    public var newFields: [AINewField]?

    enum CodingKeys: String, CodingKey {
        case title, description, priority, tags, branch, estimate, dueDate, repos, milestone, fieldValues, newFields
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        title = try container.decode(String.self, forKey: .title)
        description = try container.decodeIfPresent(String.self, forKey: .description)
        priority = try container.decode(IssuePriority.self, forKey: .priority)
        tags = try container.decodeIfPresent([String].self, forKey: .tags) ?? []
        branch = try container.decodeIfPresent(String.self, forKey: .branch)
        estimate = try container.decodeIfPresent(String.self, forKey: .estimate)
        dueDate = try container.decodeIfPresent(String.self, forKey: .dueDate)
        repos = try container.decodeIfPresent([String].self, forKey: .repos)
        milestone = try container.decodeIfPresent(String.self, forKey: .milestone)
        fieldValues = try container.decodeIfPresent([AIFieldValue].self, forKey: .fieldValues)
        newFields = try container.decodeIfPresent([AINewField].self, forKey: .newFields)
    }

    public init(
        title: String, description: String? = nil, priority: IssuePriority, tags: [String] = [],
        branch: String? = nil, estimate: String? = nil, dueDate: String? = nil, repos: [String]? = nil,
        milestone: String? = nil, fieldValues: [AIFieldValue]? = nil, newFields: [AINewField]? = nil
    ) {
        self.title = title
        self.description = description
        self.priority = priority
        self.tags = tags
        self.branch = branch
        self.estimate = estimate
        self.dueDate = dueDate
        self.repos = repos
        self.milestone = milestone
        self.fieldValues = fieldValues
        self.newFields = newFields
    }

    /// Re-validates the decoded shape against the same bounds
    /// `parsedTaskSchema` enforces — a schema-constrained CLI response can
    /// still violate these (e.g. an over-long title), so this is a real
    /// check, not a formality.
    public func validate() throws -> ParsedTask {
        guard (1...200).contains(title.count) else {
            throw AISchemaError.invalid("title must be 1-200 characters")
        }
        if let description, description.count > 2000 {
            throw AISchemaError.invalid("description must be at most 2000 characters")
        }
        guard tags.count <= 8, tags.allSatisfy({ (1...30).contains($0.count) }) else {
            throw AISchemaError.invalid("tags must be at most 8 items of 1-30 characters each")
        }
        if let branch, branch.count > 100 { throw AISchemaError.invalid("branch too long") }
        if let estimate, estimate.count > 20 { throw AISchemaError.invalid("estimate too long") }
        if let repos, repos.count > 5 { throw AISchemaError.invalid("at most 5 repos") }
        if let milestone, milestone.count > 150 { throw AISchemaError.invalid("milestone title too long") }
        if let fieldValues, fieldValues.count > 10 { throw AISchemaError.invalid("at most 10 field values") }
        if let newFields, newFields.count > 3 { throw AISchemaError.invalid("at most 3 new fields") }
        return self
    }
}

public enum AISchemaError: Error, CustomStringConvertible {
    case invalid(String)

    public var description: String {
        switch self {
        case .invalid(let message): return "Claude's response didn't match the expected format: \(message)."
        }
    }
}

public enum ParsedTasksValidator {
    public static func validate(_ tasks: [ParsedTask]) throws -> [ParsedTask] {
        guard (1...50).contains(tasks.count) else {
            throw AISchemaError.invalid("expected 1-50 tasks")
        }
        return try tasks.map { try $0.validate() }
    }
}

public enum GeneratedSubtasksValidator {
    public static func validate(_ subtasks: [String]) throws -> [String] {
        guard (1...10).contains(subtasks.count), subtasks.allSatisfy({ (1...200).contains($0.count) }) else {
            throw AISchemaError.invalid("expected 1-10 subtasks of 1-200 characters each")
        }
        return subtasks
    }
}
